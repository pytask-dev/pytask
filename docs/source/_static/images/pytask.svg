<?xml version="1.0" encoding="utf-8"?>
<!-- Generator: Adobe Illustrator 24.1.1, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" id="Capa_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 viewBox="0 0 438.9 438.9" style="enable-background:new 0 0 438.9 438.9;" xml:space="preserve">
<style type="text/css">
	.st0{fill:#C19C00;}
	.st1{fill:#13A10E;}
	.st2{fill:#C50F1F;}
</style>
<path d="M348,57.5h-39.7V39.7c0-5.7-6.3-8.4-12-8.4h-30.8C258.1,10.5,239.8,0,218.9,0c-20.7-0.8-39.5,11.9-46.5,31.3h-30.3
	c-5.7,0-11.5,2.6-11.5,8.4v17.8H90.9C67.4,57.8,48.1,76.3,47,99.8v299.4c0,23,20.9,39.7,43.9,39.7h257c23,0,43.9-16.7,43.9-39.7
	V99.8C390.7,76.3,371.5,57.8,348,57.5z M151.5,52.3h28.7c5-0.6,9-4.4,9.9-9.4c3.1-13.5,14.9-23.1,28.7-23.5
	c13.7,0.4,25.3,10.1,28.2,23.5c0.9,5.1,5.2,9,10.4,9.4h29.8v41.8H151.5V52.3z M371,399.2c0,11.5-11.5,18.8-23,18.8h-257
	c-11.5,0-23-7.3-23-18.8V99.8c1.1-12,11-21.2,23-21.4h39.7V105c0.6,5.9,5.6,10.2,11.5,9.9h154.1c6,0.3,11.2-4,12-9.9V78.4H348
	c12,0.2,21.9,9.5,23,21.4L371,399.2L371,399.2z"/>
<path class="st0" d="M179.2,233.6c-3.9-4.1-10.4-4.4-14.6-0.5l-33.4,31.9L117,250.3c-3.9-4.1-10.4-4.4-14.6-0.5
	c-4,4.2-4,10.9,0,15.2l21.4,21.9c1.9,2.1,4.5,3.2,7.3,3.1c2.8,0,5.4-1.2,7.3-3.1l40.8-38.7c4-3.7,4.3-10,0.6-14
	C179.6,234,179.4,233.8,179.2,233.6z"/>
<path d="M329.2,256H209c-5.8,0-10.4,4.7-10.4,10.4s4.7,10.4,10.4,10.4h120.2c5.8,0,10.4-4.7,10.4-10.4S334.9,256,329.2,256z"/>
<path class="st1" d="M179.2,150c-3.9-4.1-10.4-4.4-14.6-0.5l-33.4,31.9L117,166.7c-3.9-4.1-10.4-4.4-14.6-0.5c-4,4.2-4,10.9,0,15.2
	l21.4,21.9c1.9,2.1,4.5,3.2,7.3,3.1c2.8,0,5.4-1.2,7.3-3.1l40.8-38.7c4-3.7,4.3-10,0.6-14C179.6,150.4,179.4,150.2,179.2,150z"/>
<path d="M329.2,172.4H209c-5.8,0-10.4,4.7-10.4,10.4s4.7,10.4,10.4,10.4h120.2c5.8,0,10.4-4.7,10.4-10.4S334.9,172.4,329.2,172.4z"
	/>
<path class="st2" d="M179.2,317.2c-3.9-4.1-10.4-4.4-14.6-0.5l-33.4,31.9L117,333.9c-3.9-4.1-10.4-4.4-14.6-0.5
	c-4,4.2-4,10.9,0,15.2l21.4,21.9c1.9,2.1,4.5,3.2,7.3,3.1c2.8,0,5.4-1.2,7.3-3.1l40.8-38.7c4-3.7,4.3-10,0.6-14
	C179.6,317.6,179.4,317.4,179.2,317.2z"/>
<path d="M329.2,339.6H209c-5.8,0-10.4,4.7-10.4,10.4s4.7,10.4,10.4,10.4h120.2c5.8,0,10.4-4.7,10.4-10.4S334.9,339.6,329.2,339.6z"
	/>
</svg>
