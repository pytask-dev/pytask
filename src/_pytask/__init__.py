"""This module should not contain any imports except for the version."""
from __future__ import annotations


try:
    from ._version import version as __version__
except ImportError:  # pragma: no cover
    # broken installation, we don't even try unknown only works because we do poor mans
    # version compare
    __version__ = "unknown"


__all__ = ["__version__"]
