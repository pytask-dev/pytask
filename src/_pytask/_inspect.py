from __future__ import annotations

__all__ = ["get_annotations"]


from inspect import get_annotations
