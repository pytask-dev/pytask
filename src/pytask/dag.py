import itertools

import networkx as nx


def sort_tasks_topologically(dag):
    for node in nx.topological_sort(dag):
        if "task" in dag.nodes[node]:
            yield node


def descending_tasks(task_name, dag):
    for descendant in nx.descendants(dag, task_name):
        if "task" in dag.nodes[descendant]:
            yield descendant


def task_and_descending_tasks(task_name, dag):
    yield task_name
    yield from descending_tasks(task_name, dag)


def node_and_neigbors(dag, node):
    return itertools.chain([node], dag.neighbors(node))
