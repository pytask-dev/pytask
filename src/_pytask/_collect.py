"""This module contains helpers for :mod:`_pytask.collect`."""
from __future__ import annotations

import itertools
import uuid
from pathlib import Path
from typing import Any
from typing import Iterable

import attr
from _pytask.exceptions import NodeNotCollectedError
from _pytask.nodes import find_duplicates
from _pytask.nodes import MetaNode
from _pytask.session import Session


__all__ = ["convert_objects_to_node_dictionary"]


def convert_objects_to_node_dictionary(objects: Any, when: str) -> dict[Any, Any]:
    """Convert objects to node dictionary."""
    list_of_dicts = [_convert_to_dict(x) for x in objects]
    _check_that_names_are_not_used_multiple_times(list_of_dicts, when)
    nodes = _merge_dictionaries(list_of_dicts)
    return nodes


@attr.s(frozen=True)
class _Placeholder:
    """A placeholder to mark unspecified keys in dictionaries."""

    scalar = attr.ib(type=bool, default=False)
    id_ = attr.ib(factory=uuid.uuid4, type=uuid.UUID)


def _convert_to_dict(x: Any, first_level: bool = True) -> Any | dict[Any, Any]:
    """Convert any object to a dictionary."""
    if isinstance(x, dict):
        return {k: _convert_to_dict(v, False) for k, v in x.items()}
    elif isinstance(x, Iterable) and not isinstance(x, str):
        if first_level:
            return {
                _Placeholder(): _convert_to_dict(element, False)
                for i, element in enumerate(x)
            }
        else:
            return {i: _convert_to_dict(element, False) for i, element in enumerate(x)}
    elif first_level:
        return {_Placeholder(scalar=True): x}
    else:
        return x


def _collect_node(
    session: Session, path: Path, name: str, node: str | Path
) -> dict[str, MetaNode]:
    """Collect nodes for a task.

    Parameters
    ----------
    session : _pytask.session.Session
        The session.
    path : Path
        The path to the task whose nodes are collected.
    name : str
        The name of the task.
    nodes : Dict[str, Union[str, Path]]
        A dictionary of nodes parsed from the ``depends_on`` or ``produces`` markers.

    Returns
    -------
    Dict[str, MetaNode]
        A dictionary of node names and their paths.

    Raises
    ------
    NodeNotCollectedError
        If the node could not collected.

    """
    collected_node = session.hook.pytask_collect_node(
        session=session, path=path, node=node
    )
    if collected_node is None:
        raise NodeNotCollectedError(
            f"{node!r} cannot be parsed as a dependency or product for task "
            f"{name!r} in {path!r}."
        )

    return collected_node


def _check_that_names_are_not_used_multiple_times(
    list_of_dicts: list[dict[Any, Any]], when: str
) -> None:
    """Check that names of nodes are not assigned multiple times.

    Tuples in the list have either one or two elements. The first element in the two
    element tuples is the name and cannot occur twice.

    """
    names_with_provisional_keys = list(
        itertools.chain.from_iterable(dict_.keys() for dict_ in list_of_dicts)
    )
    names = [x for x in names_with_provisional_keys if not isinstance(x, _Placeholder)]
    duplicated = find_duplicates(names)

    if duplicated:
        raise ValueError(
            f"'@pytask.mark.{when}' has nodes with the same name: {duplicated}"
        )


def _union_of_dictionaries(dicts: list[dict[Any, Any]]) -> dict[Any, Any]:
    """Merge multiple dictionaries in one.

    Examples
    --------
    >>> a, b = {"a": 0}, {"b": 1}
    >>> _union_of_dictionaries([a, b])
    {'a': 0, 'b': 1}

    >>> a, b = {'a': 0}, {'a': 1}
    >>> _union_of_dictionaries([a, b])
    {'a': 1}

    """
    return dict(itertools.chain.from_iterable(dict_.items() for dict_ in dicts))


def _merge_dictionaries(list_of_dicts: list[dict[Any, Any]]) -> dict[Any, Any]:
    """Merge multiple dictionaries.

    The function does not perform a deep merge. It simply merges the dictionary based on
    the first level keys which are either unique names or placeholders. During the merge
    placeholders will be replaced by an incrementing integer.

    Examples
    --------
    >>> a, b = {"a": 0}, {"b": 1}
    >>> _merge_dictionaries([a, b])
    {'a': 0, 'b': 1}

    >>> a, b = {_Placeholder(): 0}, {_Placeholder(): 1}
    >>> _merge_dictionaries([a, b])
    {0: 0, 1: 1}

    """
    merged_dict = _union_of_dictionaries(list_of_dicts)

    if len(merged_dict) == 1 and isinstance(list(merged_dict)[0], _Placeholder):
        placeholder, value = list(merged_dict.items())[0]
        if placeholder.scalar:
            out = value
        else:
            out = {0: value}
    else:
        counter = itertools.count()
        out = {}
        for k, v in merged_dict.items():
            if isinstance(k, _Placeholder):
                while True:
                    possible_key = next(counter)
                    if possible_key not in merged_dict and possible_key not in out:
                        out[possible_key] = v
                        break
            else:
                out[k] = v

    return out
