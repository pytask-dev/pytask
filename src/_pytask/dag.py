"""Implement some capabilities to deal with the DAG."""
from __future__ import annotations

import itertools
from typing import Dict
from typing import Generator
from typing import Iterable
from typing import Set

import attr
import networkx as nx
from _pytask.console import format_strings_as_flat_tree
from _pytask.console import TASK_ICON
from _pytask.mark_utils import has_mark
from _pytask.nodes import Task


def descending_tasks(task_name: str, dag: nx.DiGraph) -> Generator[str, None, None]:
    """Yield only descending tasks."""
    for descendant in nx.descendants(dag, task_name):
        if "task" in dag.nodes[descendant]:
            yield descendant


def task_and_descending_tasks(
    task_name: str, dag: nx.DiGraph
) -> Generator[str, None, None]:
    """Yield task and descending tasks."""
    yield task_name
    yield from descending_tasks(task_name, dag)


def preceding_tasks(task_name: str, dag: nx.DiGraph) -> Generator[str, None, None]:
    """Yield only preceding tasks."""
    for ancestor in nx.ancestors(dag, task_name):
        if "task" in dag.nodes[ancestor]:
            yield ancestor


def task_and_preceding_tasks(
    task_name: str, dag: nx.DiGraph
) -> Generator[str, None, None]:
    """Yield task and preceding tasks."""
    yield task_name
    yield from preceding_tasks(task_name, dag)


def node_and_neighbors(dag: nx.DiGraph, node: str) -> Iterable[str]:
    """Yield node and neighbors which are first degree predecessors and successors.

    We cannot use ``dag.neighbors`` as it only considers successors as neighbors in a
    DAG.

    """
    return itertools.chain([node], dag.predecessors(node), dag.successors(node))


@attr.s
class TopologicalSorter:
    """The topological sorter class.

    This class allows to perform a topological sort

    """

    dag = attr.ib(type=nx.DiGraph)
    priorities = attr.ib(factory=dict, type=Dict[str, int])
    _dag_backup = attr.ib(default=None, type=nx.DiGraph)
    _is_prepared = attr.ib(default=False, type=bool)
    _nodes_out = attr.ib(factory=set, type=Set[str])

    @classmethod
    def from_dag(cls, dag: nx.DiGraph) -> TopologicalSorter:
        """Instantiate from a DAG."""
        if not dag.is_directed():
            raise ValueError("Only directed graphs have a topological order.")

        tasks = [
            dag.nodes[node]["task"] for node in dag.nodes if "task" in dag.nodes[node]
        ]
        priorities = _extract_priorities_from_tasks(tasks)

        task_names = {task.name for task in tasks}
        task_dict = {name: nx.ancestors(dag, name) & task_names for name in task_names}
        task_dag = nx.DiGraph(task_dict).reverse()

        return cls(task_dag, priorities, task_dag.copy())

    def prepare(self) -> None:
        """Perform some checks before creating a topological ordering."""
        try:
            nx.algorithms.cycles.find_cycle(self.dag)
        except nx.NetworkXNoCycle:
            pass
        else:
            raise ValueError("The DAG contains cycles.")

        self._is_prepared = True

    def get_ready(self, n: int = 1) -> list[str]:
        """Get up to ``n`` tasks which are ready."""
        if not self._is_prepared:
            raise ValueError("The TopologicalSorter needs to be prepared.")
        if not isinstance(n, int) or n < 1:
            raise ValueError("'n' must be an integer greater or equal than 1.")

        ready_nodes = {v for v, d in self.dag.in_degree() if d == 0} - self._nodes_out
        prioritized_nodes = sorted(
            ready_nodes, key=lambda x: self.priorities.get(x, 0)
        )[-n:]

        self._nodes_out.update(prioritized_nodes)

        return prioritized_nodes

    def is_active(self) -> bool:
        """Indicate whether there are still tasks left."""
        return bool(self.dag.nodes)

    def done(self, *nodes: str) -> None:
        """Mark some tasks as done."""
        self._nodes_out = self._nodes_out - set(nodes)
        self.dag.remove_nodes_from(nodes)

    def reset(self) -> None:
        """Reset an exhausted topological sorter."""
        self.dag = self._dag_backup.copy()
        self._is_prepared = False
        self._nodes_out = set()

    def static_order(self) -> Generator[str, None, None]:
        """Return a topological order of tasks as an iterable."""
        self.prepare()
        while self.is_active():
            new_task = self.get_ready()[0]
            yield new_task
            self.done(new_task)


def _extract_priorities_from_tasks(tasks: list[Task]) -> dict[str, int]:
    """Extract priorities from tasks.

    Priorities are set via the ``pytask.mark.try_first`` and ``pytask.mark.try_last``
    markers. We recode these markers to numeric values to sort all available by
    priorities. ``try_first`` is assigned the highest value such that it has the
    rightmost position in the list. Then, we can simply call :meth:`list.pop` on the
    list which is far more efficient than ``list.pop(0)``.

    """
    priorities = {
        task.name: {
            "try_first": has_mark(task, "try_first"),
            "try_last": has_mark(task, "try_last"),
        }
        for task in tasks
    }
    tasks_w_mixed_priorities = [
        name for name, p in priorities.items() if p["try_first"] and p["try_last"]
    ]

    if tasks_w_mixed_priorities:
        name_to_task = {task.name: task for task in tasks}
        reduced_names = [
            name_to_task[name].short_name for name in tasks_w_mixed_priorities
        ]
        text = format_strings_as_flat_tree(
            reduced_names, "Tasks with mixed priorities", TASK_ICON
        )
        raise ValueError(
            "'try_first' and 'try_last' cannot be applied on the same task. See the "
            f"following tasks for errors:\n\n{text}"
        )

    # Recode to numeric values for sorting.
    numeric_mapping = {(True, False): 1, (False, False): 0, (False, True): -1}
    numeric_priorities = {
        name: numeric_mapping[(p["try_first"], p["try_last"])]
        for name, p in priorities.items()
    }

    return numeric_priorities
