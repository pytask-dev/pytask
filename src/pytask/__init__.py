"""This module contains the main namespace for pytask."""
from __future__ import annotations

from _pytask import __version__
from _pytask.build import main
from _pytask.collect_utils import depends_on
from _pytask.collect_utils import parse_nodes
from _pytask.collect_utils import produces
from _pytask.compat import check_for_optional_program
from _pytask.compat import import_optional_dependency
from _pytask.config import hookimpl
from _pytask.console import console
from _pytask.database import db
from _pytask.exceptions import CollectionError
from _pytask.exceptions import ConfigurationError
from _pytask.exceptions import ExecutionError
from _pytask.exceptions import NodeNotCollectedError
from _pytask.exceptions import NodeNotFoundError
from _pytask.exceptions import PytaskError
from _pytask.exceptions import ResolvingDependenciesError
from _pytask.graph import build_dag
from _pytask.mark import Mark
from _pytask.mark import MARK_GEN as mark  # noqa: N811
from _pytask.mark import MarkDecorator
from _pytask.mark import MarkGenerator
from _pytask.mark_utils import get_all_marks
from _pytask.mark_utils import get_marks
from _pytask.mark_utils import has_mark
from _pytask.mark_utils import remove_marks
from _pytask.mark_utils import set_marks
from _pytask.models import CollectionMetadata
from _pytask.nodes import FilePathNode
from _pytask.nodes import MetaNode
from _pytask.nodes import Task
from _pytask.outcomes import CollectionOutcome
from _pytask.outcomes import count_outcomes
from _pytask.outcomes import Exit
from _pytask.outcomes import ExitCode
from _pytask.outcomes import Persisted
from _pytask.outcomes import Skipped
from _pytask.outcomes import SkippedAncestorFailed
from _pytask.outcomes import SkippedUnchanged
from _pytask.outcomes import TaskOutcome
from _pytask.report import CollectionReport
from _pytask.report import ExecutionReport
from _pytask.report import ResolvingDependenciesReport
from _pytask.session import Session
from _pytask.traceback import format_exception_without_traceback
from _pytask.traceback import remove_internal_traceback_frames_from_exc_info
from _pytask.traceback import remove_traceback_from_exc_info
from _pytask.traceback import render_exc_info
from _pytask.warnings_utils import parse_warning_filter
from _pytask.warnings_utils import warning_record_to_str
from _pytask.warnings_utils import WarningReport


# This import must come last, otherwise a circular import occurs.
from _pytask.cli import cli  # noreorder


__all__ = [
    "__version__",
    "build_dag",
    "check_for_optional_program",
    "cli",
    "console",
    "count_outcomes",
    "db",
    "depends_on",
    "format_exception_without_traceback",
    "get_all_marks",
    "get_first_non_none_value",
    "get_marks",
    "has_mark",
    "hookimpl",
    "import_optional_dependency",
    "main",
    "mark",
    "parse_nodes",
    "parse_warning_filter",
    "produces",
    "remove_internal_traceback_frames_from_exc_info",
    "remove_marks",
    "remove_traceback_from_exc_info",
    "render_exc_info",
    "set_marks",
    "warning_record_to_str",
    "CollectionError",
    "CollectionMetadata",
    "CollectionOutcome",
    "CollectionReport",
    "ConfigurationError",
    "ExecutionReport",
    "ExecutionError",
    "Exit",
    "ExitCode",
    "FilePathNode",
    "Mark",
    "MarkDecorator",
    "MarkGenerator",
    "MetaNode",
    "NodeNotCollectedError",
    "NodeNotFoundError",
    "Persisted",
    "PytaskError",
    "ResolvingDependenciesError",
    "ResolvingDependenciesReport",
    "Session",
    "Skipped",
    "SkippedAncestorFailed",
    "SkippedUnchanged",
    "Task",
    "TaskOutcome",
    "WarningReport",
]
