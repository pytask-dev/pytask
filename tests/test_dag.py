from __future__ import annotations

import textwrap
from pathlib import Path

import pytest
from _pytask.dag import pytask_dag_create_dag
from attrs import define
from pytask import build
from pytask import cli
from pytask import ExitCode
from pytask import NodeNotFoundError
from pytask import PathNode
from pytask import Task


@define
class Node(PathNode):
    """See https://github.com/python-attrs/attrs/issues/293 for property hack."""

    def state(self):
        if "missing" in self.name:
            raise NodeNotFoundError


@pytest.mark.unit()
def test_pytask_dag_create_dag():
    root = Path.cwd() / "src"
    task = Task(
        base_name="task_dummy",
        path=root,
        function=None,
        depends_on={
            0: Node.from_path(root / "node_1"),
            1: Node.from_path(root / "node_2"),
        },
    )

    dag = pytask_dag_create_dag([task])

    assert all(
        any(i in node for i in ("node_1", "node_2", "task")) for node in dag.nodes
    )


@pytest.mark.end_to_end()
def test_check_if_root_nodes_are_available(tmp_path, runner):
    source = """
    import pytask

    @pytask.mark.depends_on("in.txt")
    @pytask.mark.produces("out.txt")
    def task_d(produces):
        produces.write_text("1")
    """
    tmp_path.joinpath("task_d.py").write_text(textwrap.dedent(source))

    result = runner.invoke(cli, [tmp_path.as_posix()])

    assert result.exit_code == ExitCode.DAG_FAILED
    assert "Failures during resolving dependencies" in result.output

    # Ensure that node names are reduced.
    assert "Failures during resolving dependencies" in result.output
    assert "Some dependencies do not exist or are" in result.output
    assert tmp_path.joinpath("task_d.py").as_posix() + "::task_d" not in result.output
    assert "task_d.py::task_d" in result.output
    assert tmp_path.joinpath("in.txt").as_posix() not in result.output
    assert tmp_path.name + "/in.txt" in result.output


@pytest.mark.end_to_end()
def test_check_if_root_nodes_are_available_w_name(tmp_path, runner):
    source = """
    from pathlib import Path
    from typing_extensions import Annotated, Any
    from pytask import PathNode, PythonNode

    node1 = PathNode(name="input1", path=Path(__file__).parent / "in.txt")
    node2 = PythonNode(name="input2")

    def task_e(in1_: Annotated[Path, node1], in2_: Annotated[Any, node2]): ...
    """
    tmp_path.joinpath("task_e.py").write_text(textwrap.dedent(source))

    result = runner.invoke(cli, [tmp_path.as_posix()])

    assert result.exit_code == ExitCode.DAG_FAILED
    assert "Failures during resolving dependencies" in result.output

    # Ensure that node names are reduced.
    assert "Failures during resolving dependencies" in result.output
    assert "Some dependencies do not exist or are" in result.output
    assert tmp_path.joinpath("task_e.py").as_posix() + "::task_e" not in result.output
    assert "task_e.py::task_e" in result.output
    assert tmp_path.joinpath("in.txt").as_posix() not in result.output
    assert tmp_path.name + "/in.txt" in result.output


@pytest.mark.end_to_end()
def test_check_if_root_nodes_are_available_with_separate_build_folder(tmp_path, runner):
    tmp_path.joinpath("src").mkdir()
    tmp_path.joinpath("bld").mkdir()
    source = """
    import pytask

    @pytask.mark.depends_on("../bld/in.txt")
    @pytask.mark.produces("out.txt")
    def task_d(produces):
        produces.write_text("1")
    """
    tmp_path.joinpath("src", "task_d.py").write_text(textwrap.dedent(source))

    result = runner.invoke(cli, [tmp_path.joinpath("src").as_posix()])

    assert result.exit_code == ExitCode.DAG_FAILED
    assert "Failures during resolving dependencies" in result.output

    # Ensure that node names are reduced.
    assert "Failures during resolving dependencies" in result.output
    assert "Some dependencies do not exist" in result.output
    assert tmp_path.joinpath("task_d.py").as_posix() + "::task_d" not in result.output
    assert "task_d.py::task_d" in result.output
    assert tmp_path.joinpath("bld", "in.txt").as_posix() not in result.output
    assert tmp_path.name + "/bld/in.txt" in result.output


@pytest.mark.end_to_end()
def test_cycle_in_dag(tmp_path, runner):
    source = """
    import pytask

    @pytask.mark.depends_on("out_2.txt")
    @pytask.mark.produces("out_1.txt")
    def task_1(produces):
        produces.write_text("1")

    @pytask.mark.depends_on("out_1.txt")
    @pytask.mark.produces("out_2.txt")
    def task_2(produces):
        produces.write_text("2")
    """
    tmp_path.joinpath("task_module.py").write_text(textwrap.dedent(source))

    result = runner.invoke(cli, [tmp_path.as_posix()])

    assert result.exit_code == ExitCode.DAG_FAILED
    assert "Failures during resolving dependencies" in result.output
    assert "The DAG contains cycles which means a dependency" in result.output


@pytest.mark.end_to_end()
def test_two_tasks_have_the_same_product(tmp_path, runner):
    source = """
    import pytask

    @pytask.mark.produces("out.txt")
    def task_1(produces):
        produces.write_text("1")

    @pytask.mark.produces("out.txt")
    def task_2(produces):
        produces.write_text("2")
    """
    tmp_path.joinpath("task_d.py").write_text(textwrap.dedent(source))

    result = runner.invoke(cli, [tmp_path.as_posix()])

    assert result.exit_code == ExitCode.DAG_FAILED
    assert "Failures during resolving dependencies" in result.output
    assert "There are some tasks which produce the same output." in result.output

    # Ensure that nodes names are reduced.
    assert tmp_path.joinpath("task_d.py").as_posix() + "::task_1" not in result.output
    assert "task_d.py::task_1" in result.output
    assert tmp_path.joinpath("task_d.py").as_posix() + "::task_2" not in result.output
    assert "task_d.py::task_2" in result.output
    assert tmp_path.joinpath("out.txt").as_posix() not in result.output
    assert tmp_path.name + "/out.txt" in result.output


@pytest.mark.end_to_end()
def test_has_node_changed_catches_notnotfounderror(runner, tmp_path):
    """Missing nodes raise NodeNotFoundError when they do not exist and their state is
    requested."""
    source = """
    import pytask

    @pytask.mark.produces("file.txt")
    def task_example(produces):
        produces.write_text("test")
    """
    tmp_path.joinpath("task_example.py").write_text(textwrap.dedent(source))

    result = runner.invoke(cli, [tmp_path.as_posix()])
    assert result.exit_code == ExitCode.OK

    tmp_path.joinpath("file.txt").unlink()

    result = runner.invoke(cli, [tmp_path.as_posix()])
    assert result.exit_code == ExitCode.OK


def test_error_when_node_state_throws_error(runner, tmp_path):
    source = """
    from pytask import PythonNode

    def task_example(a = PythonNode(value={"a": 1}, hash=True)):
        pass
    """
    tmp_path.joinpath("task_example.py").write_text(textwrap.dedent(source))

    result = runner.invoke(cli, [tmp_path.as_posix()])
    assert result.exit_code == ExitCode.DAG_FAILED
    assert "task_example" in result.output


def test_python_nodes_are_unique(tmp_path):
    tmp_path.joinpath("a").mkdir()
    tmp_path.joinpath("a", "task_example.py").write_text("def task_example(a=1): pass")
    tmp_path.joinpath("b").mkdir()
    tmp_path.joinpath("b", "task_example.py").write_text("def task_example(a=2): pass")

    session = build(paths=tmp_path)

    assert session.exit_code == ExitCode.OK
    assert len(session.dag.nodes) == 4
