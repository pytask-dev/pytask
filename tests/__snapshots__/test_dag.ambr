# serializer version: 1
# name: test_check_if_root_nodes_are_available
  '''
  ───────────────────────────── Start pytask session ─────────────────────────────
  Platform: <platform> -- Python <version>, pytask <version>, pluggy <version>
  Root: <path>
  Collected 1 task.
  
  ──────────────────── Failures during resolving dependencies ────────────────────
  
  ResolvingDependenciesError: Some dependencies do not exist or are not produced 
  by any task. See the following tree which shows which dependencies are missing 
  for which tasks.
  
  Missing dependencies:
  └── 📄 test_check_if_root_nodes_are_a0/in.txt
      └── 📝 task_d.py::task_d
  
  
  (Hint: Your file-system is case-sensitive. Check the paths' capitalization 
  carefully.)
  
  ────────────────────────────────────────────────────────────────────────────────
  '''
# ---
# name: test_check_if_root_nodes_are_available_w_name
  '''
  ───────────────────────────── Start pytask session ─────────────────────────────
  Platform: <platform> -- Python <version>, pytask <version>, pluggy <version>
  Root: <path>
  Collected 1 task.
  
  ──────────────────── Failures during resolving dependencies ────────────────────
  
  ResolvingDependenciesError: Some dependencies do not exist or are not produced 
  by any task. See the following tree which shows which dependencies are missing 
  for which tasks.
  
  Missing dependencies:
  └── 📄 input1
      └── 📝 task_e.py::task_e
  
  
  (Hint: Your file-system is case-sensitive. Check the paths' capitalization 
  carefully.)
  
  ────────────────────────────────────────────────────────────────────────────────
  '''
# ---
# name: test_check_if_root_nodes_are_available_with_separate_build_folder
  '''
  ───────────────────────────── Start pytask session ─────────────────────────────
  Platform: <platform> -- Python <version>, pytask <version>, pluggy <version>
  Root: <path>
  Collected 1 task.
  
  ──────────────────── Failures during resolving dependencies ────────────────────
  
  ResolvingDependenciesError: Some dependencies do not exist or are not produced 
  by any task. See the following tree which shows which dependencies are missing 
  for which tasks.
  
  Missing dependencies:
  └── 📄 test_check_if_root_nodes_are_a2/bld/in.txt
      └── 📝 task_d.py::task_d
  
  
  (Hint: Your file-system is case-sensitive. Check the paths' capitalization 
  carefully.)
  
  ────────────────────────────────────────────────────────────────────────────────
  '''
# ---
# name: test_cycle_in_dag
  '''
  ───────────────────────────── Start pytask session ─────────────────────────────
  Platform: <platform> -- Python <version>, pytask <version>, pluggy <version>
  Root: <path>
  Collected 2 tasks.
  
  ──────────────────── Failures during resolving dependencies ────────────────────
  
  ResolvingDependenciesError: The DAG contains cycles which means a dependency is 
  directly or indirectly a product of the same task. See the following the path of
  nodes in the graph which forms the cycle.
  
  task_module.py::task_1
       ⬇
  test_cycle_in_dag0/out_1.txt
       ⬇
  task_module.py::task_2
       ⬇
  test_cycle_in_dag0/out_2.txt
       ⬇
  task_module.py::task_1
  
  ────────────────────────────────────────────────────────────────────────────────
  '''
# ---
# name: test_error_when_node_state_throws_error
  '''
  ───────────────────────────── Start pytask session ─────────────────────────────
  Platform: <platform> -- Python <version>, pytask <version>, pluggy <version>
  Root: <path>
  Collected 1 task.
  
  ──────────────────── Failures during resolving dependencies ────────────────────
  
  TypeError: unhashable type: 'dict'
  
  The above exception was the direct cause of the following exception:
  
  ResolvingDependenciesError: While checking whether dependency 
  'test_error_when_node_state_thr0/task_example.py::task_example::a' from task(s) 
  'task_example.py::task_example' exists, an error was raised.
  
  ────────────────────────────────────────────────────────────────────────────────
  '''
# ---
# name: test_two_tasks_have_the_same_product
  '''
  ───────────────────────────── Start pytask session ─────────────────────────────
  Platform: <platform> -- Python <version>, pytask <version>, pluggy <version>
  Root: <path>
  Collected 2 tasks.
  
  ──────────────────── Failures during resolving dependencies ────────────────────
  
  ResolvingDependenciesError: There are some tasks which produce the same output. 
  See the following tree which shows which products are produced by multiple 
  tasks.
  
  Products from multiple tasks:
  └── 📄 test_two_tasks_have_the_same_p0/out.txt
      ├── 📝 task_d.py::task_1
      └── 📝 task_d.py::task_2
  
  
  ────────────────────────────────────────────────────────────────────────────────
  '''
# ---
